src/Segwit.mo

```diff
@@ -34,39 +33,38 @@ module {
       #BECH32;
     };
 
-    let bech32Result : Text = Bech32.encode(hrp,
-      bech32Input.toArray(), encoding);
-
-    return switch (decode(hrp, bech32Result)) {
-      case (#ok (_)) {
-        #ok (bech32Result)
+    let bech32Result : Text = Bech32.encode(
+      hrp,
+      Buffer.toArray(bech32Input),
+      encoding,
+    );
+
+    return switch (decode(bech32Result)) {
+      case (#ok((decodedHrp, _))) {
+        // if the following fails, then there is a bug in decoding
+        assert hrp == decodedHrp;
+        #ok(bech32Result);
       };
-      case (#err (msg)) {
-        #err (msg)
+      case (#err(msg)) {
+        #err(msg);
       };
     };
   };
 
-  // Convert a segwit address into a Witness Program.
-  // Verifies the address contains the expected human-readable part and decodes
-  // using Bech32.
-  public func decode(hrp : Text, address : Text)
-    : Result.Result<WitnessProgram, Text> {
+  // Convert a segwit address into a numan-readable part (HRP) and a Witness Program.
+  // Decodes using Bech32.
+  public func decode(address : Text) : Result.Result<(Text, WitnessProgram), Text> {
```

Comment:
change in the function signature - previously it `assert`ed that the decoding of the `address` produces the expected HRP, which is unnecessary